proof/invariant-abstract/AARCH64/ArchVCPUDomain_AI.thy

```diff
@@ -0,0 +1,597 @@
+  "cur_vcpu_in_cur_domain s \<equiv>
+     case cur_vcpu_tcb s of
+       Some t \<Rightarrow> in_cur_domain t s
+     | None \<Rightarrow> True"
+
+lemmas cur_vcpu_in_cur_domain_defs = cur_vcpu_in_cur_domain_def cur_vcpu_tcb_def in_cur_domain_def
+
+lemma cur_vcpu_in_cur_domain_lift_strong:
+  assumes [wp]: "\<And>P. f \<lbrace>\<lambda>s. P (arm_current_vcpu (arch_state s))\<rbrace>"
+                "\<And>P. f \<lbrace>\<lambda>s. P (vcpu_tcbs_of s)\<rbrace>"
+                "\<And>P. f \<lbrace>\<lambda>s. P (cur_domain s)\<rbrace>"
+                "\<And>P t. f \<lbrace>etcb_at (\<lambda>t. P (etcb_domain t)) t\<rbrace>"
+  shows "f \<lbrace>cur_vcpu_in_cur_domain\<rbrace>"
+  unfolding cur_vcpu_in_cur_domain_defs
+  by (wp_pre, wps, wpsimp wp: valid_case_option_post_wp split: option.splits)
+
+lemma cur_vcpu_in_cur_domain_lift:
+  assumes "\<And>P. f \<lbrace>\<lambda>s. P (arm_current_vcpu (arch_state s))\<rbrace>"
+          "\<And>P. f \<lbrace>\<lambda>s. P (vcpu_tcbs_of s)\<rbrace>"
+          "\<And>P. f \<lbrace>\<lambda>s. P (cur_domain s)\<rbrace>"
+          "\<And>P. f \<lbrace>\<lambda>s. P (etcbs_of s)\<rbrace>"
+  shows "f \<lbrace>cur_vcpu_in_cur_domain\<rbrace>"
+ by (rule cur_vcpu_in_cur_domain_lift_strong assms)+
+
+\<comment> \<open>This simp could possibly be avoided if arm_current_vcpu updates were wrapped in a function
+    instead of directly using modifies.\<close>
+lemma cur_vcpu_in_cur_domain_current_vcpu_update[simp]:
+  "cur_vcpu_in_cur_domain (s\<lparr>arch_state := arch_state s\<lparr>arm_current_vcpu := Some (v, b)\<rparr>\<rparr>)
+   = (\<forall>t. vcpu_tcbs_of s v = Some t \<longrightarrow> in_cur_domain t s)"
+  "cur_vcpu_in_cur_domain (s\<lparr>arch_state := arch_state s\<lparr>arm_current_vcpu := None\<rparr>\<rparr>)"
+  by (auto simp: cur_vcpu_in_cur_domain_defs split: option.splits)
+
+\<comment> \<open>FIXME: use projections instead of this\<close>
+lemma cur_vcpu_in_cur_domain_updates[simp]:
+  "cur_vcpu_in_cur_domain (cur_thread_update f_ct s) = cur_vcpu_in_cur_domain s"
+  "cur_vcpu_in_cur_domain (machine_state_update f_ms s) = cur_vcpu_in_cur_domain s"
+  "cur_vcpu_in_cur_domain (cdt_update f_cdt s) = cur_vcpu_in_cur_domain s"
+  "cur_vcpu_in_cur_domain (cdt_list_update f_cdt_list s) = cur_vcpu_in_cur_domain s"
+  "cur_vcpu_in_cur_domain (work_units_completed_update f_work_units s) = cur_vcpu_in_cur_domain s"
+  "cur_vcpu_in_cur_domain (interrupt_states_update f_interrupt_states s) = cur_vcpu_in_cur_domain s"
+  "cur_vcpu_in_cur_domain (is_original_cap_update f_orig_cap s) = cur_vcpu_in_cur_domain s"
+  "cur_vcpu_in_cur_domain (arch_state_update (arm_asid_table_update f_asid_table) s) = cur_vcpu_in_cur_domain s"
+  "cur_vcpu_in_cur_domain (s\<lparr>arch_state := (arm_asid_table_update f_asid_table) (arch_state s)\<rparr>) = cur_vcpu_in_cur_domain s" \<comment> \<open>FIXME: previous line doesn't work for this, can it be generalised?\<close>
```

Comment:
It would depend on the underlying term. Nothing obvious springs to mind. Have you looked at it with quick_print on, maybe it shows some of the underlying structure?

Reply:
To make it explicit, I think the difference is between
```
λs. arch_state_update (λarch_state. arm_asid_table_update f_asid_table arch_state) s
```
and
```
λs. arch_state_update (λ_. arm_asid_table_update f_asid_table (arch_state s)) s
```